### DISCUSSION TITLE
impulse counter with Smart Meter Interface

### DISCUSSION THREAD

**Discussion by narf0815** [arendst/Tasmota#16439] [Q&A]
Hello!
i have installed an impulse counter ( "IMPULSGEBER-W mit S0-Schnittstelle" ) and haven't really thought about how i could read from the device. after some internet search i found the Smart Meter Interface but i'm out of my depth... i was able to enable scriping but i have no idea how to script... quoting from the device installation page: "The pulse generator with its optocoupler isolated S0 output can be connected directly to existing S0 meters."
![image](https://user-images.githubusercontent.com/9511238/188325156-36cd4822-3bae-4333-8f83-81705c88a331.png)

in the end i would like to report from my D1 mini to openhab via mqtt (as i do with several other tasmota devices).
would be very great if somebody could help me with the scripting part!

best regards
narf

**Comment by joba-1** [comment]
Hi,
from what I see here you get short 5V pulses on s0+ 
From what you write I think you want to count these pulses and publish them somewhere.
You wont need a smart meter or scripting for this. Just configure the pin you use in the "Configure Module" menu as a counter.
Tasmota will publish the count via mqtt where you can grab it.

One complication: D1 expects 3.3V pulses and you probably get 5V pulses (guessing from the supply voltage) - could destroy the D1 sooner or later.
So use a voltage divider with two resistors (values ~1K-50k with ~ 2:1 ratio should be ok) to adjust the voltage level like this:

```
5V----[10K]----+----[22K]----Gnd
               |
              S0+
```

**Reply by sfromis** [reply]
To me, it does not look not like 5V pulses, but that the pulses will pull down, requiring a pullup in on the MCU side. Possibly the internal pullup of the ESP chip is enough. While a Tasmota Counter indeed is an effective simple way of, well, counting pulses, IIRC (no experience with that) the Smart Meter Interface also supports this type of input, and has fuller calculations of energy flow.

**Reply by joba-1** [reply]
I think you are right. But reading the Impulsgeber-manual before connecting might be a good idea (is S0+ open when not pulled down? Would make sense).

**Reply by narf0815** [reply]
does this help? there's also a google translate option on the page...
https://www.stall.biz/project/intelligenter-impulsgeber-w-mit-s0-schnittstelle-zum-nachruesten-fuer-konventionelle-wasserzaehler
![image](https://user-images.githubusercontent.com/9511238/188327246-fd1f311c-7abd-4744-a2c8-d913be213ee1.png)

**Reply by sfromis** [reply]
This also indicates the requirement for some pullup, like the internal ESP one, for hooking up.
![image](https://user-images.githubusercontent.com/47082390/188327342-7a7b22d2-21c0-42fd-a8bf-c017b48f7785.png)

**Reply by narf0815** [reply]
this picture is related to a special homematic decive. what could i ask the developer of this device that could help me (help you...) to figure out how this could work?

**Reply by sfromis** [reply]
If you ignore the box to the right, which actually is unimportant, you still very clearly see the need for a pullup resistor. This is also indicated by the image you posted, with S0 pins connected to an optocoupler, thus with galvanic separation from the rest of the detection circuit. With that interface, you can have no output voltages on the S0 pins, and with S0- being ground, it can only be that S0+ can be pulled down for a pulse.

**Reply by narf0815** [reply]
ok. that means i have to put a 100k resistor between s0+ from the counter and the input pin on the tasmota device?

**Reply by barbudor** [reply]
No
S0+ direct to GPIO
Pull up resistor between S0+/GPIO and 3V3
100K is useless.
If ESP internal pull-up is not strong (~10K) enough, use an external one of 4K7 or 2K2
But ESP's internal pull-up should be enough

**Reply by sfromis** [reply]
The suggested 100K indicates that only a very weak pullup is required, like the internal ESP one set by Tasmota by default, and thus adding a 100K resistor would indeed not help anything.

**Reply by narf0815** [reply]
so i should just try without an additional resistor?

> Pull up resistor between S0+/GPIO and 3V3

3V3?

**Reply by sfromis** [reply]
Right, the ESP chip already has an internal pullup resistor which will be activated by default, unless you directly choose no-pullup. Most probably just fine in this case, without adding another resistor to the mix.

3V3 means the 3.3V power for the ESP chip, the red line in the image I posted. If it becomes relevant to add an external pullup resistor (probably not), that would be between 3.3V and the data pin.

**Reply by joba-1** [reply]
d1 mini pin D3 has a 10k pullup on board.

**Reply by joba-1** [reply]
And D4 / internal LED as well - you can probably see it blink on counts if you use this one

**Reply by sfromis** [reply]
Gpio0 (D3) not recommended for such inputs, as it can mess badly with the boot process.

**Reply by narf0815** [reply]
so D4 is the pin to go with?

**Reply by sfromis** [reply]
No. The only functional argument for that pin over others would be that it on dev boards like D1 Mini or NodeMCU has an external pullup resistor, which you (again and again) most likely have no need for. As a starting point, I'd go for one of gpio 4,5,12,13,14

Gpio2 (D4) does also have a led on those dev board, which can be an indicator of activity, but there is the risk that the counter circuit does not have the pulldown strength for this. Not saying that you should avoid the pin, just that it is not "**the** pin to go with".

**Reply by narf0815** [reply]
alright, i went with D6 and it looks like it is working as expected!
so now i have a simple counter without daily or weekly/monthly/... amount. if i'd need that i would have to use the Smart Meter Interface, right?
and another question. can i modify the actual counter value? the contact is mounted on my watermeter and it would be great if i could set the tasmota counter to the same value as the analog watermeter...

**Reply by sfromis** [reply]
Yes, the command `Counter` allows setting/inspecting the current value of a counter. Beware that **if** you want to process the value in Tasmota rules, that these have numerical handling as single precision floating point, meaning loss of precision above 6 digits. Also note that the counts since last saving of settings data will be lost in case of the Tasmota device getting power cycled or similar.

As already mentioned, the Smart Meter Interface does seem to support S0 pulses too, for enhanced processing, but don't ask *me* about details.....

**Reply by narf0815** [reply]
ok, i'll take a look on how to set the current value of a counter.
what exactly does "since last saving of settings data" mean?

**Reply by sfromis** [reply]
It means that if you save settings (due to setup changes, manual saving, scheduled restart or midnight), the counter value is also saved at that time. If it then counts for 8 hours, without a new save opportunity, a power outage will result in old value being restored when restarting. If you then think "lets save it all the time", beware that this would greatly contribute to flash chip wear diminishing chip lifetime.

**Reply by narf0815** [reply]
ok thanks! guess i'll just let openhab do the calculations!

**Reply by Ekran** [reply]
To set an counter use the command `sensor53 c0 1234` 
change `0` if you have more than one smartmeter. `1234` includes the impulses/unit. For 100 impules per unit the `1234` means 12.34 units.
Source https://tasmota.github.io/docs/Smart-Meter-Interface/#sensors
